rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,481 @@
+be used in names etc. Initially only base64 encoding will be supported for
+converting from bytes to strings.
+
+As an example, if a parameter named `user_id` is extracted for the label, a
+user-specific node can use the base64 encoding of the user id to create a unique
+name:
+
+```toml
+[node_template]
+name = "user_{base64($user_id)}"
+```
+
+The initial impementation of dynamic node creation will focus only on support
+for per-user labels, as it is the first motivating example we have. Each
+extracted parameter will be the matching bytes that represent the user identity
+(currently the HMAC of the bearer token, but in future it would be the user's
+public key).
+
+A new LabelTemplate protobuf object will be implemented to support parameterised
+labels. It will reuse the existing tags from the Label, but also add templated
+tags. These can be used for matching labels and extracting parameters, as well
+as generating concrete label instances given a parameter value.
+
+```proto
+
+// Template for generating labels given parameter values, or matching labels
+// and extracting parameter values.
+message LabelTemplate {
+  repeated TagTemplate confidentiality_tags = 1;
+  repeated TagTemplate integrity_tags = 2;
+}
+
+// Template for generating a user-specific gRPC tag given a user ID, or for
+// extracting a user ID parameter from a matched gRPC tag.
+message GrpcTagTemplate {
+  string user_id_parameter_name = 1;
+}
+
+
+// Templated tags for supporting templated labels.
+message TagTemplate {
+  oneof tag {
+    GrpcTag grpc_tag = 1;
+    GrpcTagTemplate grpc_tag_template = 1;
+    WebAssemblyModuleTag web_assembly_module_tag = 2;
+    WebAssemblyModuleSignatureTag web_assembly_module_signature_tag = 3;
+    TlsEndpointTag tls_endpoint_tag = 4;
+  }
+}
+
+```
+
+As an example, the following can be used to define a label template matching a
+user-specific gRPC confidentiality tag and extracting the byte-array content of
+the tag as a parameter named user_id:
+
+```toml
+
+[[label_template.confidentiality_tags]]
+[label_template.confidentiality_tags.tag.grpc_tag_template]
+user_id_parameter_name = "user_id"
+
+```
+
+Note: I am unsure about the exact TOML representation for this.
+
+Question: In general TOML is not great for deeply nested object hierarchies, so
+we should perhaps consider a different serialisation format. Prost does not
+support text-serialised protos. JSON is an option, but not that great for human
+readable config. YAML can handle the complex object hierarchies, but the
+specification is very complex and parsing YAML is considered risky due to
+potential of injection vulnerabilities that is common when trying to support
+dynamic local object creation as defined by the specification. The application
+configuration is in TOML format, so having to deal with multiple serialisation
+formats when creating an application is not a great developer experience. Is
+there a suitable format that will work for everything?
```

Comment:
1) I'm not sure if we need nested object hierarchies, unless we are going to create connected nodes as substructures of nodes that create them.

2) About choosing the format, I think YAML better fits the picture, because it can handle complex structures (it is used in production for defining large services in Kubernetes), TOML may become incomprehensible after a certain point, and JSON is incomprehensible from the start =)

3) And I'm still leaning towards having this functionality implemented as a global `init` function (where we create nodes with `node_create` and connect them with `connect(src, dst, channel_name)`), rather than defining it in a separate file, because we will give developers the full expressiveness of Rust for creating their complex graphs.

Reply:
> * I'm not sure if we need nested object hierarchies, unless we are going to create connected nodes as substructures of nodes that create them.

There is already quite a deep object hierarchy that is difficult to represent with TOML, with a template containing a list of node templates. A node template contains a label template. A label template contains tag tempaltes etc. 

> * About choosing the format, I think YAML better fits the picture, because it can handle complex structures (it is used in production for defining large services in Kubernetes), TOML may become incomprehensible after a certain point, and JSON is incomprehensible from the start =)

Perhap something like StrictYAML will remove the security risk while keeping the main structure of YAML.

> * And I'm still leaning towards having this functionality implemented as a global `init` function (where we create nodes with `node_create` and connect them with `connect(src, dst, channel_name)`), rather than defining it in a separate file, because we will give developers the full expressiveness of Rust for creating their complex graphs.

One of the main reasons for this RFC is to support analysis of node graphs to find potential problems. Problems with the node graph will be a huge source of bugs and security risks for developers, so analysis tools would be a requirement in my opinion. If the creation is done in Rust code without a manifest file we would need to do symbolic execution of the code to determine all the possible node graphs that could be created and analyse each one of those. This would be much harder to do, and might never complete.